drivers/display/ssd1306.c

```diff
@@ -48,144 +49,130 @@ LOG_MODULE_REGISTER(ssd1306, CONFIG_DISPLAY_LOG_LEVEL);
 
 struct ssd1306_data {
 	const struct device *reset;
-	const struct device *i2c;
+	const struct device *bus;
+#if DT_INST_ON_BUS(0, spi)
+	struct spi_cs_control cs_ctrl;
+	struct spi_config spi_config;
+	const struct device *data_cmd;
+#endif
 	uint8_t contrast;
 	uint8_t scan_mode;
 };
 
-static inline int ssd1306_reg_read(struct ssd1306_data *driver,
-				   uint8_t reg, uint8_t * const val)
+static inline int ssd1306_write_bus(const struct device *dev,
+				    uint8_t *buf, size_t len, bool command)
 {
-	return i2c_reg_read_byte(driver->i2c,
-				 DT_INST_REG_ADDR(0),
-				 reg, val);
-}
+	struct ssd1306_data *driver = dev->data;
+#if DT_INST_ON_BUS(0, i2c)
+	return i2c_burst_write(driver->bus, DT_INST_REG_ADDR(0),
+			       command ? SSD1306_CONTROL_ALL_BYTES_CMD :
+					 SSD1306_CONTROL_ALL_BYTES_DATA,
+			       buf, len);
+#elif DT_INST_ON_BUS(0, spi)
+	int errno;
+
+	gpio_pin_set(driver->data_cmd, DT_INST_GPIO_PIN(0, data_cmd_gpios),
+		     command ? 0 : 1);
+	struct spi_buf tx_buf = {
+		.buf = buf,
+		.len = len
+	};
 
-static inline int ssd1306_reg_write(struct ssd1306_data *driver,
-				    uint8_t reg, uint8_t val)
-{
-	return i2c_reg_write_byte(driver->i2c,
-				  DT_INST_REG_ADDR(0),
-				  reg, val);
-}
+	struct spi_buf_set tx_bufs = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
 
-static inline int ssd1306_reg_update(struct ssd1306_data *driver, uint8_t reg,
-				     uint8_t mask, uint8_t val)
-{
-	return i2c_reg_update_byte(driver->i2c,
-				   DT_INST_REG_ADDR(0),
-				   reg, mask, val);
+	errno = spi_write(driver->bus, &driver->spi_config, &tx_bufs);
+	gpio_pin_set(driver->data_cmd,
+		     DT_INST_GPIO_PIN(0, data_cmd_gpios), 0);
+
+	return errno;
+#endif
 }
 
 static inline int ssd1306_set_panel_orientation(const struct device *dev)
 {
-	struct ssd1306_data *driver = dev->data;
 	uint8_t cmd_buf[] = {
-		SSD1306_CONTROL_BYTE_CMD,
```

Comment:
Please split your PR in few commits and do this optimization (and related [changes](https://github.com/zephyrproject-rtos/zephyr/pull/28322/commits/a53c308e1458fadf31e8a2561a6ddaf3e5143bf6#diff-5fd46386a4f473a41f96f54a41f497e3R11)) as a first step.

Reply:
For now I just splitted the PR Into 2 commits, within the current PR:
- optimize protocol
- add SPI support
If requested, I can also add more splits or split the PR itself.